### PR TITLE
Convert a `let` to `const` for a variable declaration

### PR DESCRIPTION
**Pull request by nickebbitt**
Implements #107 

Doesn't currently convert where there are multiple variable declarators on the same line e.g. 

```
let someVariable, otherVariable = 'value';
someVariable = 'anotherValue';
```

This would require splitting the declarations which is doable but I felt maybe an iteration once we have a working baseline.

### CODE REVIEW

**Comment by nickebbitt** [comment]
@nicoespeon interested in your feedback on the impl and can you think of any edge cases I've not covered that we'd like tests for?

File: src/refactorings/convert-let-to-const/convert-let-to-const.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import { Editor, Code, ErrorReason } from "../../editor/editor";
+import { Selection } from "../../editor/selection";
+import * as t from "../../ast";
+import { isLet } from "../../ast";
+import { Binding } from "@babel/traverse";
```
Comment (nickebbitt) [inline src/refactorings/convert-let-to-const/convert-let-to-const.ts:5]: @nicoespeon I'm currently importing `@babel/traverse` directly here however was wondering if maybe we'd prefer this behaviour to live in the `ast` module?

File: src/refactorings/convert-let-to-const/convert-let-to-const.test.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  let otherVariable = 'otherValue';
+}`,
+        expected: `{
+  const someVariable = 'someValue';
+  let otherVariable = 'otherValue';
+}`
+      },
+      {
+        description:
+          "multiple variables delcared seperately, other one mutated",
```
[the remark this answers is not in the document]
Comment (nickebbitt) [reply src/refactorings/convert-let-to-const/convert-let-to-const.test.ts:54]: yes that makes sense, my bad, I'd interpreted the selection to have a wider scope (i.e. all variables on that line) than it does, rather than actually a single selected variable.

I agree with your thoughts so will make the changes 👍

File: src/refactorings/convert-let-to-const/convert-let-to-const.test.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        description: "non-mutated variable declared in a block",
+        code: `{
+  let someVariable = 'value';
+}`,
+        expected: `{
+  const someVariable = 'value';
+}`
+      },
+      {
+        description: "multiple non-mutated variables declared together",
```
[the remark this answers is not in the document]
Comment (nickebbitt) [reply src/refactorings/convert-let-to-const/convert-let-to-const.test.ts:33]: Thinking about this one, I kind of overlooked the fact that the refactor would leave us with invalid code due to the `'const' declarations must be initialized.` error.

At least initially I'm going to prevent the refactor for this case.

Eventually, I think the solution for this is to move the selected variable onto its own line leaving the other variable unaffected by the refactor.

@nicoespeon does that feel like a reasonable decision?

File: src/refactorings/convert-let-to-const/convert-let-to-const.ts (hunk 1 above)
Comment (nickebbitt) [reply src/refactorings/convert-let-to-const/convert-let-to-const.ts:4]: @nicoespeon I think this is the only change left to make, I looked at the `ast` module but unsure where this should go.

Does it fit in one of the existing files or require a new one?

Comment (nicoespeon) [reply src/refactorings/convert-let-to-const/convert-let-to-const.ts:4]: It has to do with `@babel/traverse` so I'd expose it here: https://github.com/nicoespeon/abracadabra/blob/3326ba97cfe1a8049f741148fc12c19ed308729a/src/ast/transformation.ts#L2

File: src/refactorings/convert-let-to-const/convert-let-to-const.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (nicoespeon) [reply src/refactorings/convert-let-to-const/convert-let-to-const.test.ts]: Yep, I think it's the best decision. We shouldn't turn an un-initialized `let` into `const` or that would be invalid. Of course it makes the multi-variables trickier to deal with, but it's the right thing to do 👍

File: src/refactorings/convert-let-to-const/convert-let-to-const.ts (hunk 1 above)
Comment (nickebbitt) [reply src/refactorings/convert-let-to-const/convert-let-to-const.ts]: Done 👍 

Unless you've any other feedback this hopefully we're good to merge :)

**Comment by nickebbitt** [comment]
@nicoespeon I added some details to the README, not sure if the section is exactly right? https://github.com/nicoespeon/abracadabra/pull/109/commits/ea5c9592b73ffe28111b69740bd3f482f1902537

**Comment by nicoespeon** [comment]
@nickebbitt I think the section is just right 👌 

I'll have a final look tomorrow but that looks great to me already 👏

File: src/refactorings/convert-let-to-const/convert-let-to-const.test.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        code: `{
+  let someVariable = 'value';
+}`,
+        selection: Selection.cursorAt(1, 7),
+        expected: `{
+  const someVariable = 'value';
+}`
+      },
+      {
+        description: "only the selected non-mutated variable",
```
Comment (nicoespeon) [inline src/refactorings/convert-let-to-const/convert-let-to-const.test.ts:35]: 👏

File: src/refactorings/convert-let-to-const/convert-let-to-const.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+function createVisitor(
+  selection: Selection,
+  onMatch: (path: t.NodePath<t.VariableDeclarator>) => void
+): t.Visitor {
+  return {
+    VariableDeclarator(path) {
+      const { node } = path;
+      if (!selection.isInsideNode(node)) {
```
Comment (nicoespeon) [inline src/refactorings/convert-let-to-const/convert-let-to-const.ts:47]: 👍

File: src/refactorings/convert-let-to-const/convert-let-to-const.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  };
+}
+
+function isSingleLetVariableDeclaration(
+  node: t.Node
+): node is t.VariableDeclaration {
+  return t.isLet(node) && node.kind === "let" && node.declarations.length === 1;
+}
+
+function variableCanBeConst(
```
Comment (nicoespeon) [inline src/refactorings/convert-let-to-const/convert-let-to-const.ts:61]: 👍
